python/private/platform_runtime_repo.bzl

```diff
@@ -0,0 +1,196 @@
+load(":repo_utils.bzl", "REPO_DEBUG_ENV_VAR", "repo_utils")
+
+_TOOLCHAIN_IMPL_TEMPLATE = """\
+# Generated by python/private/platform_runtime_repo.bzl
+
+load("@rules_python//python:py_runtime.bzl", "py_runtime")
+load("@rules_python//python:py_runtime_pair.bzl", "py_runtime_pair")
+load("@rules_python//python/cc:py_cc_toolchain.bzl", "py_cc_toolchain")
+load("@rules_python//python/private:py_exec_tools_toolchain.bzl", "py_exec_tools_toolchain")
+
+cc_library(
+    name = "python_headers",
+    srcs = glob(["include/**/*.h"]),
+    includes = ["include"],
+)
+
+cc_library(
+    name = "libpython",
+    # Don't use a recursive glob because the lib/ directory usually contains
+    # a subdirectory of the stdlib -- lots of unrelated files
+    srcs = glob([
+        "lib/*{lib_ext}",  # Match libpython*.so
+        "lib/*{lib_ext}*",  # Also match libpython*.so.1.0
+    ]),
+    hdrs = [":python_headers"],
+)
+
+py_runtime(
+    name = "py3_runtime",
+    interpreter_path = "{interpreter_path}",
+    python_version = "PY3",
+    interpreter_version_info = {{
+        "major": "{major}",
+        "minor": "{minor}",
+        "micro": "{micro}",
+    }},
+    implementation_name = "{implementation_name}",
+)
+
+py_runtime_pair(
+    name = "python_runtimes",
+    py2_runtime = None,
+    py3_runtime = ":py3_runtime",
+)
+
+py_exec_tools_toolchain(
+    name = "py_exec_tools_toolchain",
+)
+
+py_cc_toolchain(
+    name = "py_cc_toolchain",
+    headers = ":python_headers",
+    libs = ":libpython",
+    python_version = "{major}.{minor}.{micro}",
+)
+"""
+
+_GET_INFO_PROGRAM = """
+import sys
+import sysconfig
+import json
+
+data = {
+  "major": sys.version_info.major,
+  "minor": sys.version_info.minor,
+  "micro": sys.version_info.micro,
+  "include": sysconfig.get_path("include"),
+  "implementation_name": sys.implementation.name,
+}
+
+config_vars = [
+    "LDLIBRARY", "LIBDIR", "INSTSONAME", "LIBDEST", "PY3LIBRARY",
+    "SHLIB_SUFFIX",
+]
+data.update(zip(config_vars, sysconfig.get_config_vars(*config_vars)))
+print(json.dumps(data))
+"""
+
+def _platform_runtime_repo_impl(rctx):
+    interpreter_path = resolve_interpreter_path(rctx, rctx.attr)
+    if not interpreter_path or not interpreter_path.exists:
+        fail("Unable to find interpreter from {}".format(rctx.attr.interpreter_path))
+
+    rctx.file("getinfo.py", _GET_INFO_PROGRAM)
+    output = repo_utils.execute_checked_stdout(
+        rctx,
+        op = "platform_runtime_repo.GetPythonInfo({})".format(rctx.name),
+        arguments = [
+            interpreter_path,
+            rctx.path("getinfo.py"),
+        ],
+        quiet = True,
+    )
+    rctx.delete("getinfo.py")
+    info = json.decode(output)
+
+    # The cc_library.includes values have to be non-absolute paths, otherwise
+    # the toolchain will give an error. Work around this error by making them
+    # appear as part of this repo.
+    rctx.symlink(info["include"], "include")
```

Comment:
Yeah the symlink creation is good, but note that on Windows it will do a copy if the symlinks are not enabled. Maybe worth putting into a comment.

That said, this only works on non-windows, I guess.

Reply:
A copy would be fine. It's copying the whole directory, right?

Reply:
I think so.